File: Yashwanth_24b1039/week4/2_smallest_divisor.py
class Solution(object):
    def is_possible(self,nums,threshold,divisor):

        sum=0
        for x in nums:
            sum+=(x+divisor-1)//divisor
        
        if (sum>threshold):
            return True 
        return False
        
    def smallestDivisor(self, nums, threshold):
        """
        :type nums: List[int]
        :type threshold: int
        :rtype: int
        """
        sum=0

        for x in nums:
            sum+=x
        
        low=1
        high=max(nums)
        mid=0

        while(high>=low):

            mid=(high+low)//2
            if(not self.is_possible(nums,threshold,mid)):
                high=mid-1
            else:
                low=mid+1
            
        return low

File: Karan_24b3003/Week 4/2_smallest_divisor.cpp
class Solution {
public:
    int computeSum(vector<int>& nums, int divisor) {
        int sum = 0;
        for (int num : nums) {
            sum += (num + divisor - 1) / divisor;
        }
        return sum;
    }

    int smallestDivisor(vector<int>& nums, int threshold) {
        int left = 1;
        int right = *max_element(nums.begin(), nums.end());
        int result = right;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int currSum = computeSum(nums, mid);

            if (currSum <= threshold) {
                result = mid;
                right = mid - 1;
            } else
                left = mid + 1;
        }
        return result;
    }
};
